Add a keyboard shortcut to toggle hidden event visibility when labs are enabled

UwUnyaa opened the pull request:
This is an initial PR implementing changes requested here: https://github.com/vector-im/element-web/issues/20646

The keyboard shortcut as of now is control (or cmd) shift h. I don't have means to check whether this conflicts with anything on macOS.

Since the setting is only available when labs are enabled, the shortcut is also only added when that setting is toggled. Because of the nature of the setting, actual handling of it is implemented in LoggedInView, but it might make sense to move that somewhere else, as there haven't been any labs-specific shortcuts before. Most of the changes in this PR are nothing but plumbing a new section with keybindings, with a tiny bit of extra code to make sure it won't run when labs are not enabled.

Notes: The keyboard shortcut is control (or cmd) shift h.

Type: T-Enhancement

<!-- CHANGELOG_PREVIEW_START -->
---
Here's what your changelog entry will look like:

## ✨ Features
 * The keyboard shortcut is control (or cmd) shift h. ([\#7584](https://github.com/matrix-org/matrix-react-sdk/pull/7584)). Contributed by @UwUnyaa.<!-- CHANGELOG_PREVIEW_END -->

UwUnyaa commented:
I don't think the bot picks up edits that add labels, which is likely why checks are failing as of now.

HarHarLinks commented:
@ element team: while reviewing, please also comment on two ideas i raised in the original issue:
1. Does the setting "Show hidden events in timeline" really belong in labs? "Developer mode" is in labs as well, but because it is unstable. While I would consider viewing "hidden events" and advanced feature for enthusiasts, it isn't only for developers working on element, but simply with element on something else. You might be using element to debug a bot/appservice sending replies, or set up something like hookshot or populus that use custom events. Is showing of "hidden events" unstable in that sense?
2. What about adding this same toggle to the new quick settings? Needs design review I suppose?

SimonBrandner commented on src/accessibility/KeyboardShortcuts.ts:
```diff
@@ -276,8 +278,22 @@ export const shortcuts: Record<Categories, IShortcut[]> = {
 };
 
+// Add this short only if labs are enabled, as the feature is only exposed there.
+if (SdkConfig.get()['showLabsSettings']) {
+    shortcuts[Categories.LABS].push({
+        keybinds: [{
+            modifiers: [CMD_OR_CTRL, Modifiers.SHIFT],
+            key: Key.H,
+        }],
+        description: _td("Toggle visibility of hidden events"),
+    });
+}
```
This should probably use `registerShortcut()` and be somewhere else, I think.

UwUnyaa replied:
What would be the good place for that? I can't find any calls to that function anywhere within code.

UwUnyaa replied:
From my poking around, `SdkConfig` is not initialized when `shortcuts` are being defined, so the key shortcuts will have to be defined somewhere else, although I can't find any use of the `registerShortcut` function, so I'm left guessing.

SimonBrandner replied:
Perhaps a call from https://github.com/vector-im/element-web/blob/develop/src/vector/index.ts to a function in https://github.com/vector-im/element-web/blob/develop/src/vector/init.tsx at the correct time would work?

Though someone more experienced might have a better idea

dbkr replied:
Ah, registerShortcut is used from the element-web in the electron-specific code: https://github.com/vector-im/element-web/blob/develop/src/vector/platform/ElectronPlatform.tsx#L290

And yep, Simon's suggestion seems like a sensible place (`registerShortcuts` maybe).

dbkr replied:
You could also presumably add the shortcut anyway and just do nothing if labs aren't shown.

UwUnyaa replied:
@dbkr actually, that might be a good idea. The list of categories of shortcuts is defined in the shortcuts tab, and I managed to add a check there to append the Labs section to the list. Having the shortcut defined unconditionally in one place as everything else might make things easier to follow.

SimonBrandner commented on src/KeyBindingsDefaults.ts:
```diff
@@ -411,10 +412,24 @@ const navigationBindings = (): KeyBinding<NavigationAction>[] => {
 };
 
+const labsBindings = (): KeyBinding<LabsAction>[] => {
+    return [
+        {
+            action: LabsAction.ToggleHiddenEventVisibility,
+            keyCombo: {
+                key: Key.H,
+                ctrlOrCmd: true,
+                shiftKey: true,
+            },
+        },
+    ];
```
I believe this should return an empty array, if `SdkConfig.get()['showLabsSettings']` is false

SimonBrandner commented on src/components/structures/LoggedInView.tsx:
```diff
@@ -531,6 +533,33 @@ class LoggedInView extends React.Component<IProps, IState> {
                 handled = PlatformPeg.get().onKeyDown(ev);
         }
+
+        // Handle labs actions here, as they apply within the same scope
+        if (SdkConfig.get()['showLabsSettings'] && !handled) {
```
I believe we should check for `SdkConfig.get()['showLabsSettings']` in `src/KeyBindingsDefaults.ts` instead

UwUnyaa replied:
`SdkConfig` is not initialized yet when that code runs.

SimonBrandner replied:
Really? From what I remember that would only run when a keybinding is actually pressed, though I might be wrong

UwUnyaa replied:
From my testing, `SdkConfig` hasn't loaded any configuration when `KeyBindingsDefaults` is ran, so for labs visibility just yields `undefined`. Should I create a separate file for labs keybinds, and use `registerShortcut()` within it?

dbkr replied:
Yeah, looks like KeyBindingsDeafults is all just done at import-time. I think this is probably the correct place as far as I can see.

SimonBrandner replied:
Hmm, ok, I don't feel super happy about this we can figure out a way to improve upon this in the future

_Resolved_

UwUnyaa replied:
Ok, the check in `KeyBindingsDefaults` works, but it doesn't work in `KeyboardShortcuts`. I've moved the check there, and the shortcuts tab in settings checks whether labs shortcuts should be displayed at all.

SimonBrandner reviewed (changes requested):
Otherwise looks good - thanks!

SimonBrandner reviewed (approved):
LGTM now - thanks for working on this! Will make my life so much easier!

SimonBrandner commented:
@UwUnyaa, I have caused some merge conflicts by https://github.com/matrix-org/matrix-react-sdk/pull/7562, sorry! Could you please have a look into resolving them?

SimonBrandner commented:
> @UwUnyaa, I have caused some merge conflicts by #7562, sorry! Could you please have a look into resolving them?

(I can have a look into it myself, if necessary, as I caused the merge conflicts)

UwUnyaa commented:
@SimonBrandner I've rebased this PR, and fixed things up. I had to add logic of displaying the labs section in a different way now, but that's just a couple extra lines of code.

SimonBrandner commented on src/components/views/settings/tabs/user/KeyboardUserSettingsTab.tsx:
```diff
@@ -76,6 +77,12 @@ interface IKeyboardShortcutRowProps {
     name: string;
 }
 
+// Filter out the labs section if labs aren't enabled.
+const visibleCategories = SdkConfig.get()['showLabsSettings']
+    ? Object.entries(CATEGORIES)
+    : Object.entries(CATEGORIES)
+        .filter(([categoryName]) => categoryName !== CategoryName.LABS);
```
I think I'd prefer something like `Object.entries(CATEGORIES).filter([categoryName] => categoryName !== CategoryName.LABS || SdkConfig.get()['showLabsSettings'])` but that is probably just a personal preference

SimonBrandner reviewed (approved):
Thanks!

PS: Could you please merge instead of rebase next time? Rebasing can confuse GH and make reviewing a bit harder, so we tend to prefer merges
